README.md
# @ipoa/utils

[![npm version](https://img.shields.io/npm/v/@ipoa/utils.svg?style=flat-square)](https://www.npmjs.org/package/@ipoa/utils)
[![plugin version](https://shields.io/badge/@ipoa/utils-v0.0.7-blue)](https://www.npmjs.org/package/@ipoa/utils)
[![install size](https://packagephobia.com/badge?p=@ipoa/utils)](https://packagephobia.com/result?p=@ipoa/utils)

## install

```shell
npm install @ipoa/utils --save
```

## Example

```javascript
// Determine if a value is an Array
var utils = require('@ipoa/utils')
console.log(utils.isArray([1, 2, 3])) // true

// merges the properties of each object and returns result. 
var result = utils.merge({ foo1: 123 }, { foo1: 456 }, { foo2: 789 })
console.log(result); // outputs { foo1: 456, foo2: 789 }

utils.extend(utils, {
  /**
   * Compose the `User-Agent` content
   *
   * @returns {string} - The `User-Agent` value
   */
  userAgent: function () {
    const { name: pkgName, version: pkgVersion } = require('../package.json')
    const { platform, arch, versions: { node: nodeVersion } } = process

    return `${pkgName}/${pkgVersion} node/${nodeVersion} ${platform}/${arch}`
  },
})
module.exports = utils

// getUrlParams
var url = 'http://xxx.xxx.com/?id=1&start=0&count=10'
var result = utils.getUrlParams(url);
console.log(result); // outputs { id: '1', start: '0', count: '10' }

```

## Features

- Determine if a value is undefined:`isUndefined`
- Determine if a value is a Buffer:`isBuffer`
- Determine if a value is an ArrayBuffer:`isArrayBuffer`
- Determine if a value is a FormData:`isFormData`
- Determine if a value is a view on an ArrayBuffer:`isArrayBufferView`
- Determine if a value is a String:`isString`
- Determine if a value is a Number:`isNumber`
- Determine if a value is an Object:`isObject`
- Determine if a value is a plain Object:`isPlainObject`
- Determine if a value is an JSON:`isJSON`
- Determine if a value is a Date:`isDate`
- Determine if a value is a File:`isFile`
- Determine if a value is a Blob:`isBlob`
- Determine if a value is a Function:`isFunction`
- Determine if a value is a Stream:`isStream`
- Determine if a value is a URLSearchParams object:`isURLSearchParams`
- Trim excess whitespace off the beginning and end of a string:`trim`
- Determine if we're running in a standard browser environment:`isProcessEnv`
- Similar to `isStandardBrowserEnv`, just check it's running in a Node environment
- Determine if a value is a `form-data` node module
- Iterate over an Array or an Object invoking a function for each item:`forEach`
- merges the properties of each object and returns result:`merge`
- Extends object a by mutably adding to it the properties of object b:`extend`
- Remove byte order marker. This catches EF BB BF (the UTF-8 BOM):`stripBOM`
- Parse parameters from URL:`getUrlParams`

## Unit Testing

To run the tests for should simply run:

```shell
npm install && npm test
```

## Resources

- [changelog](./README.md)

## License

[MIT](./LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.6](https://github.com/ipoa/node-utils/compare/v0.0.5...v0.0.6) (2021-03-28)


### Features

* Parse parameters from URL ([07be400](https://github.com/ipoa/node-utils/commit/07be400752b5280fdbfbd31cc8b049c7ab33b158))

### [0.0.5](https://github.com/ipoa/node-utils/compare/v0.0.4...v0.0.5) (2021-03-18)


### Features

* Determine if a value is an JSON:`isJSON`  node module ([4dd7a06](https://github.com/ipoa/node-utils/commit/4dd7a06c802151a9a540211ed3130c319aaf17ff))

### 0.0.2 (2021-03-13)

### Features

* add utils function ([85cdff2](https://github.com/ipoa/node-utils/commit/85cdff20982936a9ca7c64732146d09f1d454cab))


tests/lib/utils.test.js
var should = require('should')
var utils = require('../../lib/utils')

describe('lib/utils', () => {
  it('should be an Object', () => {
    utils.should.be.an.Object()
  })

  it('should have `isString` property and be a function', () => {
    utils.should.have.property('isString')
    should(utils.isString).be.a.Function()
  })

  it('should have `isNumber` property and be a function', () => {
    utils.should.have.property('isNumber')
    should(utils.isNumber).be.a.Function()
  })

  it('should have `isBuffer` property and be a function', () => {
    utils.should.have.property('isBuffer')
    should(utils.isBuffer).be.a.Function()
  })

  it('should have `isObject` property and be a function', () => {
    utils.should.have.property('isObject')
    should(utils.isObject).be.a.Function()
  })

  describe('utils::isJSON', () => {
    it('should have `isJSON` property and be a function', () => {
      utils.should.have.property('isJSON')
      should(utils.isJSON).be.a.Function()
    })

    it('method `isJSON` returns the result test', function () {
      var goodJson = '{"a":"obja","b":[0,1,2],"c":{"d":"some object"}}'
      var badJson = '{"a":"obja""b":[0,1,2],"c":{"d":"some object"}}'
      var strNumber = '121212'
      should(utils.isJSON(goodJson)).eql(true)
      should(utils.isJSON(badJson)).eql(true)
      should(utils.isJSON(strNumber)).eql(false)
      // check is an object
      var object = { a: 12, b: [1, 2, 3] }
      should(utils.isJSON(object)).eql(false)
      should(utils.isJSON(object, true)).eql(true)
    })
  })

  it('should have `extend` property and be a function', () => {
    utils.should.have.property('extend')
    should(utils.extend).be.a.Function()
  })

  describe('utils::isProcessEnv', () => {
    it('method `isProcessEnv` should be function', () => {
      should(utils.isProcessEnv).be.a.Function()
    })

    it('method `isProcessEnv` should returns Boolean', () => {
      utils.isProcessEnv().should.be.a.Boolean()
    })

    it('method `isProcessEnv` should returns boolean `true` under Node environment', () => {
      utils.isProcessEnv().should.be.a.Boolean().and.be.True()
    })
  })

  describe('utils::isProcessFormData', () => {
    it('method `isProcessFormData` should be function', () => {
      should(utils.isProcessFormData).be.a.Function()
    })

    it('method `isProcessFormData` should returns Boolean', () => {
      utils.isProcessFormData().should.be.a.Boolean()
    })

    const mockup = {
      toString: () => '[object FormData]'
    }
    it('method `isProcessFormData` should returns boolean `true` under Node environment', () => {
      mockup.should.be.an.Object()
      utils.isProcessFormData(mockup).should.be.a.Boolean().and.be.True()
    })
  })
  describe('utils::getUrlParams', () => {
    it('method `getUrlParams` should be function', () => {
      utils.getUrlParams.should.be.a.Function()
    })

    it('method `getUrlParams` should returns object', () => {
      utils.getUrlParams().should.be.a.Object()
    })

    it('method `getUrlParams` loads url parameters  should  return  object result', () => {
      const url = 'http://xxx.xxx.com/?id=1&start=0&count=10'
      should(utils.getUrlParams(url)).eql({ id: '1', start: '0', count: '10' })
    })
  })
})


lib/utils.d.ts
export interface Utils {

    /**
     * Determine if a value is an Array
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is an Array, otherwise false
     */
    isArray(val): boolean

    /**
     * Determine if a value is undefined
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if the value is undefined, otherwise false
     */
    isUndefined(val): boolean

    /**
     * Determine if a value is a Buffer
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a Buffer, otherwise false
     */
    isBuffer(val): boolean

    /**
     * Determine if a value is an ArrayBuffer
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is an ArrayBuffer, otherwise false
     */
    isArrayBuffer(val): boolean

    /**
     * Determine if a value is a FormData
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is an FormData, otherwise false
     */

    isFormData(val): boolean

    /**
     * Determine if a value is a view on an ArrayBuffer
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a view on an ArrayBuffer, otherwise false
     */

    isArrayBufferView(val): boolean

    /**
     * Determine if a value is a String
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a String, otherwise false
     */

    isString(val): boolean

    /**
     * Determine if a value is a Number
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a Number, otherwise false
     */

    isNumber(val): boolean

    /**
     * Determine if a value is an Object
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is an Object, otherwise false
     */

    isObject(val): boolean

    /**
     * Determine if a value is a plain Object
     *
     * @param {Object} val The value to test
     * @return {boolean} True if value is a plain Object, otherwise false
     */

    isPlainObject(val): boolean

    /**
     * Determine if a value is a Date
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a Date, otherwise false
     */

    isDate(val): boolean

    /**
     * Determine if a value is a File
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a File, otherwise false
     */
    isFile(val): boolean

    /**
     * Determine if a value is a Blob
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a Blob, otherwise false
     */

    isBlob(val): boolean

    /**
     * Determine if a value is a Function
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a Function, otherwise false
     */

    isFunction(val): boolean

    /**
     * Determine if a value is a Stream
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a Stream, otherwise false
     */

    isStream(val): boolean

    /**
     * Determine if a value is a URLSearchParams object
     *
     * @param {Object} val The value to test
     * @returns {boolean} True if value is a URLSearchParams object, otherwise false
     */

    isURLSearchParams(val): boolean

    /**
     * Trim excess whitespace off the beginning and end of a string
     *
     * @param {String} str The String to trim
     * @returns {String} The String freed of excess whitespace
     */

    trim(str: string): boolean

    /**
     * Determine if we're running in a standard browser environment
     *
     * This allows axios to run in a web worker, and react-native.
     * Both environments support XMLHttpRequest, but not fully standard globals.
     *
     * web workers:
     *  typeof window -> undefined
     *  typeof document -> undefined
     *
     * react-native:
     *  navigator.product -> 'ReactNative'
     * nativescript
     *  navigator.product -> 'NativeScript' or 'NS'
     */
    isStandardBrowserEnv(): boolean

    /**
     * Similar to `isStandardBrowserEnv`, just check it's running in a Node environment
     *
     * @return {boolean} Ture on Node, otherwise false
     */
    isProcessEnv(): boolean

    /**
     * Determine if a value is a `form-data` node module
     *
     *
     * @param {Object} val - To test value
     *
     * @returns {boolean} True if value is a `form-data` module, otherwise false
     */
    isProcessFormData(val): boolean

    /**
     * Iterate over an Array or an Object invoking a function for each item.
     *
     * If `obj` is an Array callback will be called passing
     * the value, index, and complete array for each item.
     *
     * If 'obj' is an Object callback will be called passing
     * the value, key, and complete object for each property.
     *
     * @param {Object|Array} obj The object to iterate
     * @param {Function} fn The callback to invoke for each item
     */

    forEach(obj, fn): void

    /**
     * Accepts varargs expecting each argument to be an object, then
     * immutably merges the properties of each object and returns result.
     *
     * When multiple objects contain the same key the later object in
     * the arguments list will take precedence.
     *
     * Example:
     *
     * ```js
     * var result = merge({foo: 123}, {foo: 456});
     * console.log(result.foo); // outputs 456
     * ```
     * merge( obj1, obj2, obj3, ... )
     *
     * @param args
     * @returns {Object} Result of all merge properties
     */

    merge(...args: any[]): object

    /**
     * Extends object a by mutably adding to it the properties of object b.
     *
     * @param {Object} a The object to be extended
     * @param {Object} b The object to copy properties from
     * @param {Object} thisArg The object to bind function to
     * @return {Object} The resulting value of object a
     */
    extend<T>(a, b: T, thisArg?: any): T & this

    /**
     * Remove byte order marker. This catches EF BB BF (the UTF-8 BOM)
     *
     * @param {string} content with BOM
     * @return {string} content value without BOM
     */

    stripBOM(content): string

    /**
     * @param {String} url
     * @description Parse parameters from URL
     *
     * Example:
     * ```js
     * var url = 'http://xxx.xxx.com/?id=1&start=0&count=10'
     * var result = getUrlParams(url);
     * console.log(result); // outputs { id: '1', start: '0', count: '10' }
     * ```
     * @return {Object} parameters as Object
     *
     */
    getUrlParams(url: string): object

    [key: string]: any

}

declare const utils: Utils
export default utils


lib/utils.js
'use strict'

var bind = require('./helpers/bind')

// utils is a library of generic helper functions non-specific to ipoa

var toString = Object.prototype.toString

/**
 * Determine if a value is an Array
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is an Array, otherwise false
 */
function isArray(val) {
  return toString.call(val) === '[object Array]'
}

/**
 * Determine if a value is undefined
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if the value is undefined, otherwise false
 */
function isUndefined(val) {
  return typeof val === 'undefined'
}

/**
 * Determine if a value is a Buffer
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is a Buffer, otherwise false
 */
function isBuffer(val) {
  return val !== null && !isUndefined(val) && val.constructor !== null && !isUndefined(val.constructor)
    && typeof val.constructor.isBuffer === 'function' && val.constructor.isBuffer(val)
}

/**
 * Determine if a value is an ArrayBuffer
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is an ArrayBuffer, otherwise false
 */
function isArrayBuffer(val) {
  return toString.call(val) === '[object ArrayBuffer]'
}

/**
 * Determine if a value is a FormData
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is an FormData, otherwise false
 */
function isFormData(val) {
  return (typeof FormData !== 'undefined') && (val instanceof FormData)
}

/**
 * Determine if a value is a view on an ArrayBuffer
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is a view on an ArrayBuffer, otherwise false
 */
function isArrayBufferView(val) {
  var result
  if ((typeof ArrayBuffer !== 'undefined') && (ArrayBuffer.isView)) {
    result = ArrayBuffer.isView(val)
  } else {
    result = (val) && (val.buffer) && (val.buffer instanceof ArrayBuffer)
  }
  return result
}

/**
 * Determine if a value is a String
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is a String, otherwise false
 */
function isString(val) {
  return typeof val === 'string'
}

/**
 * Determine if a value is a Number
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is a Number, otherwise false
 */
function isNumber(val) {
  return typeof val === 'number'
}

/**
 * Determine if a value is an Object
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is an Object, otherwise false
 */
function isObject(val) {
  return val !== null && typeof val === 'object'
}

/**
 * Determine if a value is a plain Object
 *
 * @param {Object} val The value to test
 * @return {boolean} True if value is a plain Object, otherwise false
 */
function isPlainObject(val) {
  if (toString.call(val) !== '[object Object]') {
    return false
  }

  var prototype = Object.getPrototypeOf(val)
  return prototype === null || prototype === Object.prototype
}

/**
 * Determine if a value is an JSON
 *
 * @param str
 * @param passObject
 * @returns {boolean} True if value is an JSON, otherwise false
 */
function isJSON(str, passObject = false) {
  if (isObject(str) && passObject) return true

  if (!isString(str)) return false

  str = str.replace(/\s/g, '').replace(/\n|\r/, '')

  if (/^\{(.*?)\}$/.test(str))
    return /"(.*?)":(.*?)/g.test(str)

  if (/^\[(.*?)\]$/.test(str)) {
    return str.replace(/^\[/, '')
      .replace(/\]$/, '')
      .replace(/},{/g, '}\n{')
      .split(/\n/)
      .map(function (s) { return isJSON(s) })
      .reduce(function (prev, curr) { return !!curr })
  }

  return false
}

/**
 *
 * Determine if a value is a Date
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is a Date, otherwise false
 */
function isDate(val) {
  return toString.call(val) === '[object Date]'
}

/**
 * Determine if a value is a File
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is a File, otherwise false
 */
function isFile(val) {
  return toString.call(val) === '[object File]'
}

/**
 * Determine if a value is a Blob
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is a Blob, otherwise false
 */
function isBlob(val) {
  return toString.call(val) === '[object Blob]'
}

/**
 * Determine if a value is a Function
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is a Function, otherwise false
 */
function isFunction(val) {
  return toString.call(val) === '[object Function]'
}

/**
 * Determine if a value is a Stream
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is a Stream, otherwise false
 */
function isStream(val) {
  return isObject(val) && isFunction(val.pipe)
}

/**
 * Determine if a value is a URLSearchParams object
 *
 * @param {Object} val The value to test
 * @returns {boolean} True if value is a URLSearchParams object, otherwise false
 */
function isURLSearchParams(val) {
  return typeof URLSearchParams !== 'undefined' && val instanceof URLSearchParams
}

/**
 * Trim excess whitespace off the beginning and end of a string
 *
 * @param {String} str The String to trim
 * @returns {String} The String freed of excess whitespace
 */
function trim(str) {
  return str.replace(/^\s*/, '').replace(/\s*$/, '')
}

/**
 * Determine if we're running in a standard browser environment
 *
 * This allows axios to run in a web worker, and react-native.
 * Both environments support XMLHttpRequest, but not fully standard globals.
 *
 * web workers:
 *  typeof window -> undefined
 *  typeof document -> undefined
 *
 * react-native:
 *  navigator.product -> 'ReactNative'
 * nativescript
 *  navigator.product -> 'NativeScript' or 'NS'
 */
function isStandardBrowserEnv() {
  if (typeof navigator !== 'undefined' && (navigator.product === 'ReactNative' ||
    navigator.product === 'NativeScript' ||
    navigator.product === 'NS')) {
    return false
  }
  return (
    typeof window !== 'undefined' &&
    typeof document !== 'undefined'
  )
}

/**
 * Similar to `isStandardBrowserEnv`, just check it's running in a Node environment
 *
 * @return {boolean} Ture on Node, otherwise false
 */
function isProcessEnv() {
  return typeof process !== 'undefined' && toString.call(process) === '[object process]'
}

/**
 * Determine if a value is a `form-data` node module
 *
 *
 * @param {Object} val - To test value
 *
 * @returns {boolean} True if value is a `form-data` module, otherwise false
 */
function isProcessFormData(val) {
  // toString.call(val) returns `[object Object]`, let's checking the instance `toString`
  return this.isProcessEnv() && this.isObject(val) && val.toString() === '[object FormData]'
}

/**
 * Iterate over an Array or an Object invoking a function for each item.
 *
 * If `obj` is an Array callback will be called passing
 * the value, index, and complete array for each item.
 *
 * If 'obj' is an Object callback will be called passing
 * the value, key, and complete object for each property.
 *
 * @param {Object|Array} obj The object to iterate
 * @param {Function} fn The callback to invoke for each item
 */
function forEach(obj, fn) {
  // Don't bother if no value provided
  if (obj === null || typeof obj === 'undefined') {
    return
  }

  // Force an array if not already something iterable
  if (typeof obj !== 'object') {
    /*eslint no-param-reassign:0*/
    obj = [obj]
  }

  if (isArray(obj)) {
    // Iterate over array values
    for (var i = 0, l = obj.length; i < l; i++) {
      fn.call(null, obj[i], i, obj)
    }
  } else {
    // Iterate over object keys
    for (var key in obj) {
      if (Object.prototype.hasOwnProperty.call(obj, key)) {
        fn.call(null, obj[key], key, obj)
      }
    }
  }
}

/**
 * Accepts varargs expecting each argument to be an object, then
 * immutably merges the properties of each object and returns result.
 *
 * When multiple objects contain the same key the later object in
 * the arguments list will take precedence.
 *
 * Example:
 *
 * ```js
 * var result = merge({foo: 123}, {foo: 456});
 * console.log(result.foo); // outputs 456
 * ```
 *
 * @param {Object} obj1 Object to merge
 * @returns {Object} Result of all merge properties
 */
function merge(/* obj1, obj2, obj3, ... */) {
  var result = {}

  function assignValue(val, key) {
    if (isPlainObject(result[key]) && isPlainObject(val)) {
      result[key] = merge(result[key], val)
    } else if (isPlainObject(val)) {
      result[key] = merge({}, val)
    } else if (isArray(val)) {
      result[key] = val.slice()
    } else {
      result[key] = val
    }
  }

  for (var i = 0, l = arguments.length; i < l; i++) {
    forEach(arguments[i], assignValue)
  }
  return result
}

/**
 * Extends object a by mutably adding to it the properties of object b.
 *
 * @param {Object} a The object to be extended
 * @param {Object} b The object to copy properties from
 * @param {Object} thisArg The object to bind function to
 * @return {Object} The resulting value of object a
 */
function extend(a, b, thisArg) {
  forEach(b, function assignValue(val, key) {
    if (thisArg && typeof val === 'function') {
      a[key] = bind(val, thisArg)
    } else {
      a[key] = val
    }
  })
  return a
}

/**
 * Remove byte order marker. This catches EF BB BF (the UTF-8 BOM)
 *
 * @param {string} content with BOM
 * @return {string} content value without BOM
 */
function stripBOM(content) {
  if (content.charCodeAt(0) === 0xFEFF) {
    content = content.slice(1)
  }
  return content
}

/**
 * @param {String} url
 * @description Parse parameters from URL
 *
 * Example:
 * ```js
 * var url = 'http://xxx.xxx.com/?id=1&start=0&count=10'
 * var result = getUrlParams(url);
 * console.log(result); // outputs { id: '1', start: '0', count: '10' }
 * ```
 * @return {Object} parameters as Object
 *
 */

function getUrlParams(url = '') {
  var keyValueArr = url && url.split('?')[1].split('&')
  var paramObj = {}
  forEach(keyValueArr, function (item) {
    var keyValue = item.split('=')
    paramObj[keyValue[0]] = keyValue[1]
  })
  return paramObj
}

module.exports = {
  isArray: isArray,
  isArrayBuffer: isArrayBuffer,
  isBuffer: isBuffer,
  isFormData: isFormData,
  isArrayBufferView: isArrayBufferView,
  isString: isString,
  isNumber: isNumber,
  isObject: isObject,
  isPlainObject: isPlainObject,
  isJSON: isJSON,
  isUndefined: isUndefined,
  isDate: isDate,
  isFile: isFile,
  isBlob: isBlob,
  isFunction: isFunction,
  isStream: isStream,
  isURLSearchParams: isURLSearchParams,
  isStandardBrowserEnv: isStandardBrowserEnv,
  isProcessEnv: isProcessEnv,
  isProcessFormData: isProcessFormData,
  forEach: forEach,
  merge: merge,
  extend: extend,
  trim: trim,
  stripBOM: stripBOM,
  getUrlParams: getUrlParams
}
